### PR TITLE
BottomSheet UX for verification

### LINKED ISSUE #709: Verifications in DMS / Handle outgoing transactions
- [ ] Modify profile screen to start
- [x] Bottom sheet for SAS verification

### LINKED ISSUE #732: SDK | MSC2366 Key verification flow addition: m.key.verification.ready
The current key verification framework is asymmetrical in that the user who requests the verification is unable to select the key verification method. This makes it harder for more experienced users who wish to guide less experienced users through the verification process, especially if they are not verifying in-person, but are using a trusted but remote channel of verification (such as telephone or video conference).

https://github.com/uhoreg/matrix-doc/blob/key_verification_accept/proposals/2366-key-verification-accept.md

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes #732
Fixes #709

verif

### CODE REVIEW

File: vector/src/main/res/layout/fragment_verification_conclusion.xml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    <ImageView
+        android:id="@+id/verifyConclusionImageView"
+        android:layout_width="180dp"
+        android:layout_height="180dp"
+        android:layout_gravity="center_horizontal"
+        android:layout_marginTop="16dp"
+        tools:background="@drawable/ic_shield_trusted"
+        android:background="@drawable/ic_shield_trusted"
+        android:visibility="visible"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_conclusion.xml]: can be removed

File: vector/src/main/res/layout/fragment_verification_conclusion.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        android:textColor="?riotx_text_secondary"
+        app:layout_constraintTop_toBottomOf="@id/verificationConclusionTitle"/>
+
+    <ImageView
+        android:id="@+id/verifyConclusionImageView"
+        android:layout_width="180dp"
+        android:layout_height="180dp"
+        android:layout_gravity="center_horizontal"
+        android:layout_marginTop="16dp"
+        tools:background="@drawable/ic_shield_trusted"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_conclusion.xml]: can be removed, also, attribute are not sorted well, please format all new layout files with attributes sorting on.

File: vector/src/main/res/layout/fragment_verification_conclusion.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/verifyConclusionDescription" />
+
+    <TextView
+        android:id="@+id/verifyConclusionBottomDescription"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="8dp"
+        tools:text="@string/verification_green_shield"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_conclusion.xml]: remove

File: vector/src/main/res/layout/fragment_verification_request.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:orientation="vertical"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_request.xml]: remove

File: vector/src/main/res/layout/fragment_verification_request.xml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    <!--    app:layout_constraintTop_toBottomOf="@id/verificationRequestAvatar"-->
+
+    <com.google.android.material.button.MaterialButton
+        android:id="@+id/verificationStartButton"
+        style="@style/VectorButtonStylePositive"
+        android:layout_width="match_parent"
+        android:layout_marginTop="16dp"
+        android:text="@string/start_verification"
+        app:layout_constraintTop_toBottomOf="@id/verificationRequestText"
+        tools:visibility="visible" />
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_request.xml]: remove

File: vector/src/main/res/layout/fragment_verification_request.xml [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+    <TextView
+        android:id="@+id/verificationWaitingText"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center_horizontal"
+        android:textColor="?vctr_notice_secondary"
+        android:textSize="17sp"
+        android:textStyle="bold"
+        android:visibility="invisible"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_verification_request.xml]: Add `tools:visibility="visible"` please

File: vector/src/main/res/values/strings_riotX.xml [hunk 7]
```diff
@@ -5,15 +5,25 @@
...
+
     <string name="notification_initial_sync">Initial Sync…</string>
 
-    <string name="sent_a_file">File</string>
-    <string name="sent_an_audio_file">Audio</string>
-    <string name="sent_an_image">Image.</string>
-    <string name="sent_a_video">Video.</string>
+    <string name="verification_conclusion_warning">Untrusted sign in</string>
+    <string name="verification_sas_match">They match</string>
+    <string name="verification_sas_do_not_match">They don‘t match</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings_riotX.xml]: Strange apostrophe

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt [hunk 8]
```diff
@@ -39,6 +40,10 @@ interface SasVerificationService {
 
     fun getExistingTransaction(otherUser: String, tid: String): SasVerificationTransaction?
 
+    fun getExistingVerificationRequest(otherUser: String): List<PendingVerificationRequest>?
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt]: done

File: matrix-sdk-android/build.gradle [hunk 9]
```diff
@@ -106,6 +106,7 @@ dependencies {
 
     implementation "androidx.lifecycle:lifecycle-extensions:$lifecycle_version"
     kapt "androidx.lifecycle:lifecycle-compiler:$lifecycle_version"
+    implementation("com.google.guava:guava:28.2-jre")
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/build.gradle]: done. It's not used anymore

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt [hunk 10]
```diff
@@ -59,11 +66,29 @@ interface SasVerificationService {
...
+    }
+
+    companion object {
+
+        fun isValidRequest(age: Long?): Boolean {
+            if (age == null) return false
+            val now = System.currentTimeMillis()
+            val tooInThePast = now - (10 * 60 * 1000)
+            val fiveMinInMs = 5 * 60 * 1000
+            val tooInTheFuture = System.currentTimeMillis() + fiveMinInMs
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt [hunk 11]
```diff
@@ -59,11 +66,29 @@ interface SasVerificationService {
...
+        fun verificationRequestUpdated(pr: PendingVerificationRequest) {}
+    }
+
+    companion object {
+
+        fun isValidRequest(age: Long?): Boolean {
+            if (age == null) return false
+            val now = System.currentTimeMillis()
+            val tooInThePast = now - (10 * 60 * 1000)
+            val fiveMinInMs = 5 * 60 * 1000
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt]: k

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationTransaction.kt [hunk 12]
```diff
@@ -47,4 +47,6 @@ interface SasVerificationTransaction {
      * both short codes do match
      */
     fun userHasVerifiedShortCode()
+
+    fun shortCodeDoNotMatch()
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationTransaction.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt [hunk 13]
```diff
@@ -59,11 +66,29 @@ interface SasVerificationService {
...
+            val tooInTheFuture = System.currentTimeMillis() + fiveMinInMs
+
+            return !(age < tooInThePast || age > tooInTheFuture)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/crypto/sas/SasVerificationService.kt]: > return age in tooInThePast..tooInTheFuture

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/message/MessageVerificationDoneContent.kt [hunk 14]
```diff
@@ -17,12 +17,20 @@ package im.vector.matrix.android.api.session.room.model.message
...
         @Json(name = "m.relates_to") val relatesTo: RelationDefaultContent?
 ) : VerificationInfo {
-    override fun isValid() = true
+
+    override val transactionID: String?
+        get() = relatesTo?.eventId
+
+    override fun isValid() = transactionID?.isNotEmpty() == true
+
+    override fun toEventContent(): Content? = this.toContent()
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/message/MessageVerificationDoneContent.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/message/MessageVerificationReadyContent.kt [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+internal data class MessageVerificationReadyContent(
+        @Json(name = "from_device") override val fromDevice: String? = null,
+        @Json(name = "methods") override val methods: List<String>? = null,
+        @Json(name = "m.relates_to") val relatesTo: RelationDefaultContent?
+) : VerificationInfoReady {
+
+    override val transactionID: String?
+        get() = relatesTo?.eventId
+
+    override fun toEventContent() = this.toContent()
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/api/session/room/model/message/MessageVerificationReadyContent.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/CryptoModule.kt [hunk 16]
```diff
@@ -180,6 +180,9 @@ internal abstract class CryptoModule {
     @Binds
     abstract fun bindEncryptEventTask(encryptEventTask: DefaultEncryptEventTask): EncryptEventTask
 
+    @Binds
+    abstract fun bindSendVerificationMessageTask(sendDefaultSendVerificationMessageTask: DefaultSendVerificationMessageTask): SendVerificationMessageTask
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/CryptoModule.kt]: noted

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/tasks/RoomVerificationUpdateTask.kt [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+import im.vector.matrix.android.internal.task.Task
+import timber.log.Timber
+import java.util.*
+import javax.inject.Inject
+
+internal interface RoomVerificationUpdateTask : Task<RoomVerificationUpdateTask.Params, Unit> {
+    data class Params(
+            val events: List<Event>,
+            val sasVerificationService: DefaultSasVerificationService,
+            val cryptoService: CryptoService
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/tasks/RoomVerificationUpdateTask.kt]: No choice for now, or we have a dependency cycle (Sas Service is injected in CryptoService) so cannot inject cryptoService in task

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SasTransportRoomMessage.kt [hunk 18]
```diff
@@ -167,16 +256,39 @@ internal class SasTransportRoomMessage(
...
 internal class SasTransportRoomMessageFactory @Inject constructor(
-        private val sendVerificationMessageTask: DefaultSendVerificationMessageTask,
-        private val taskExecutor: TaskExecutor) {
+        private val monarchy: Monarchy,
+        private val localEchoEventFactory: LocalEchoEventFactory) {
 
-    fun createTransport(roomId: String,
-                        cryptoService: CryptoService,
-                        tx: SASVerificationTransaction?
+    fun createTransport(context: Context, userId: String, userDevice: String, roomId: String, tx: SASVerificationTransaction?
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SasTransportRoomMessage.kt]: Done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SendVerificationMessageWorker.kt [hunk 19]
```diff
@@ -0,0 +1,66 @@
+package im.vector.matrix.android.internal.crypto.verification
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SendVerificationMessageWorker.kt]: done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SendVerificationMessageWorker.kt [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        } catch (exception: Throwable) {
+            if (exception.shouldBeRetried()) {
+                Result.retry()
+            } else {
+                Result.success(errorOutputData)
+            }
+        }
+    }
+
+    private fun Throwable.shouldBeRetried(): Boolean {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/crypto/verification/SendVerificationMessageWorker.kt]: Done

File: matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/DefaultRoomService.kt [hunk 21]
```diff
@@ -71,6 +71,20 @@ internal class DefaultRoomService @Inject constructor(private val monarchy: Mona
         }
     }
 
+    override fun getExistingDirectRoomWithUser(otherUserId: String): Room? {
+        Realm.getInstance(monarchy.realmConfiguration).use { realm ->
+            val roomId = RoomSummaryEntity.where(realm)
+                    .equalTo(RoomSummaryEntityFields.IS_DIRECT, true)
+                    .findAll()?.let { dms ->
+                        dms.firstOrNull {
+                            it.otherMemberIds.contains(otherUserId)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/DefaultRoomService.kt]: It's not used right now (missing user profile page to plug it);
Thx for the hint, will check when profile page will be there

[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/session/room/DefaultRoomService.kt]: idem

File: matrix-sdk-android/src/main/res/values/strings_RiotX.xml [hunk 22]
```diff
@@ -21,4 +21,21 @@
...
+    <string name="verify_open_camera_link">open your camera</string>
+
+    <string name="verify_by_emoji_title">Verify by Emoji</string>
+    <string name="verify_by_emoji_description">If you can’t scan the code above, verify by comparing a short, unique selection of emoji.</string>
+
+    <string name="aria_qr_code_description">QR code image</string>
+
+    <string name="verification_request_alert_title">Verify %s</string>
+    <string name="verification_request_waiting_for">Waiting for %s…</string>
+    <string name="verification_request_alert_description">For extra security, verify %s by checking a one-time code on both your devices.\n\nFor maximum security, do this in person.</string>
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply matrix-sdk-android/src/main/res/values/strings_RiotX.xml]: good catch, thx

File: vector/src/main/java/im/vector/riotx/core/utils/SpannableUtils.kt [hunk 23]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2018 New Vector Ltd
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/core/utils/SpannableUtils.kt]: :D

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/SASVerificationCodeViewModel.kt [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+        val isWaitingFromOther: Boolean = false
+) : MvRxState
+
+class SASVerificationCodeViewModel @AssistedInject constructor(
+        @Assisted initialState: SASVerificationCodeViewState,
+        private val session: Session
+) : VectorViewModel<SASVerificationCodeViewState, EmptyAction>(initialState), SasVerificationService.SasVerificationListener {
+
+    init {
+        withState { state ->
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/SASVerificationCodeViewModel.kt]: yes, forgot to do it for this class

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/SASVerificationCodeFragment.kt [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+import kotlinx.android.synthetic.main.fragment_bottom_sas_verification_code.*
+import javax.inject.Inject
+
+class SASVerificationCodeFragment @Inject constructor(
+        val viewModelFactory: SASVerificationCodeViewModel.Factory
+) : VectorBaseFragment() {
+
+    override fun getLayoutResId() = R.layout.fragment_bottom_sas_verification_code
+
+    @BindView(R.id.sas_emoji_grid)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/SASVerificationCodeFragment.kt]: Yes this class is deleted in next pr

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheet.kt [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+
+    @BindView(R.id.verificationRequestName)
+    lateinit var otherUserNameText: TextView
+
+    @BindView(R.id.verificationRequestAvatar)
+    lateinit var otherUserAvatarImageView: ImageView
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        val view = inflater.inflate(R.layout.bottom_sheet_verification, container, false)
+        ButterKnife.bind(this, view)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheet.kt]: Ok done for this one, but will need to be done for all bottomsheet, extract layour id etc.. Another PR?

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheet.kt [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+            }
+        })
+    }
+
+    override fun invalidate() = withState(viewModel) {
+        it.otherUserMxItem?.let { matrixItem ->
+            val displayName = matrixItem.displayName ?: ""
+            otherUserNameText.text = getString(R.string.verification_request_alert_title, displayName)
+                    .toSpannable()
+                    .colorizeMatchingText(displayName, ThemeUtils.getColor(requireContext(), R.attr.vctr_notice_text_color))
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheet.kt]: I think we can come with something better for simple formatting on string resources, let's move to another PR?

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheetViewModel.kt [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+                ?: return@withState
+        when (action) {
+            is VerificationAction.RequestVerificationByDM -> {
+//                session
+                setState {
+                    copy(pendingRequest = session.getSasVerificationService().requestKeyVerificationInDMs(otherUserId, roomId))
+                }
+            }
+            is VerificationAction.StartSASVerification    -> {
+                val request = session.getSasVerificationService().getExistingVerificationRequest(otherUserId)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationBottomSheetViewModel.kt]: it does exist, missed this refactoring

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationChooseMethodViewModel.kt [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+        val SASMOdeAvailable: Boolean = false
+) : MvRxState
+
+class VerificationChooseMethodViewModel @AssistedInject constructor(
+        @Assisted initialState: VerificationChooseMethodViewState,
+        private val session: Session
+) : VectorViewModel<VerificationChooseMethodViewState, EmptyAction>(initialState) {
+
+    init {
+        withState { state ->
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationChooseMethodViewModel.kt]: done

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationConclusionFragment.kt [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+        when (it.conclusionState) {
+            ConclusionState.SUCCESS   -> {
+                verificationConclusionTitle.text = getString(R.string.sas_verified)
+                verifyConclusionDescription.setTextOrHide(getString(R.string.sas_verified_successful_description))
+                verifyConclusionBottomDescription.text = getString(R.string.verification_green_shield)
+                verifyConclusionImageView.setImageDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.ic_shield_trusted))
+            }
+            ConclusionState.WARNING   -> {
+                verificationConclusionTitle.text = getString(R.string.verification_conclusion_not_secure)
+                verifyConclusionDescription.setTextOrHide(null)
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationConclusionFragment.kt]: done

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationRequestViewModel.kt [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+import im.vector.matrix.android.api.session.crypto.sas.SasVerificationTransaction
+import im.vector.matrix.android.api.util.MatrixItem
+import im.vector.matrix.android.api.util.toMatrixItem
+import im.vector.matrix.android.internal.crypto.verification.PendingVerificationRequest
+import im.vector.riotx.core.di.HasScreenInjector
+import im.vector.riotx.core.platform.VectorViewModel
+
+data class VerificationRequestViewState(
+        val roomId: String? = null,
+        val matrixItem: MatrixItem,
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationRequestViewModel.kt]: done

File: vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationRequestViewModel.kt [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+                        started = Success(false).takeIf { pr == null }
+                                ?: Success(true).takeIf { pr?.isReady == true }
+                                ?: Loading(),
+                        matrixItem = it.toMatrixItem()
+                )
+            }
+        }
+    }
+
+    override fun handle(action: VerificationAction) {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/crypto/verification/VerificationRequestViewModel.kt]: Done

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt [hunk 33]
```diff
@@ -1114,7 +1131,8 @@ class RoomDetailFragment @Inject constructor(
     }
 
     override fun onAvatarClicked(informationData: MessageInformationData) {
-        vectorBaseActivity.notImplemented("Click on user avatar")
+        // vectorBaseActivity.notImplemented("Click on user avatar")
+        roomDetailViewModel.handle(RoomDetailAction.RequestVerification(informationData.senderId))
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt]: yes, there is an issue to fix it

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt [hunk 34]
```diff
@@ -960,6 +961,22 @@ class RoomDetailFragment @Inject constructor(
                             }
                         }
                     }
+                    is RoomDetailAction.RequestVerification       -> {
+                        VerificationBottomSheet().apply {
+                            arguments = Bundle().apply {
+                                putParcelable(MvRx.KEY_ARG, VerificationBottomSheet.VerificationArgs(data.userId, roomId = roomDetailArgs.roomId))
+                            }
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/RoomDetailFragment.kt]: done

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/VerificationRequestItem.kt [hunk 35]
```diff
@@ -101,13 +102,16 @@ abstract class VerificationRequestItem : AbsBaseMessageItem<VerificationRequestI
                 }
                 holder.statusTextView.isVisible = true
             }
-            else                                -> {
+            else -> {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/VerificationRequestItem.kt]: my new Android studio had default formatting rules

File: vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/VerificationRequestItem.kt [hunk 36]
```diff
@@ -101,13 +102,16 @@ abstract class VerificationRequestItem : AbsBaseMessageItem<VerificationRequestI
...
                 holder.buttonBar.isVisible = false
                 holder.statusTextView.text = null
                 holder.statusTextView.isVisible = false
             }
         }
 
+        // Always hide buttons if request is too old
+        holder.buttonBar.isVisible = holder.buttonBar.isVisible && SasVerificationService.isValidRequest(attributes.informationData.ageLocalTS)
+
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/riotx/features/home/room/detail/timeline/item/VerificationRequestItem.kt]: Done

File: vector/src/main/res/layout/fragment_bottom_sas_verification_code.xml [hunk 37]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ScrollView xmlns:android="http://schemas.android.com/apk/res/android"
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_bottom_sas_verification_code.xml]: yes, bad copy paste

File: vector/src/main/res/layout/fragment_verification_conclusion.xml (hunk 1 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_conclusion.xml]: Done

File: vector/src/main/res/layout/fragment_verification_conclusion.xml (hunk 2 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_conclusion.xml]: Done

File: vector/src/main/res/layout/fragment_verification_request.xml (hunk 4 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_request.xml]: done

File: vector/src/main/res/layout/fragment_verification_request.xml (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_request.xml]: done

File: vector/src/main/res/layout/fragment_verification_request.xml (hunk 6 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_request.xml]: done

File: vector/src/main/res/values/strings_riotX.xml (hunk 7 above)
Comment (BillCarsonFr) [reply vector/src/main/res/values/strings_riotX.xml]: done

File: vector/src/main/res/layout/fragment_verification_conclusion.xml (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/fragment_verification_conclusion.xml]: done

**Pull request merged**